File: fastify-ts-bun-backend/src/errors/error-map.ts
import { ZodError } from 'zod';
import { AppError } from './app-error';
import { InternalServerError, RateLimitExceededError, ValidationError } from './http-errors';

type FastifyJsonError = Error & {
  statusCode?: number;
  code?: string;
};

function isMalformedJsonError(error: unknown): error is FastifyJsonError {
  if (!(error instanceof Error)) {
    return false;
  }

  const candidate = error as FastifyJsonError;
  return candidate.code === 'FST_ERR_CTP_INVALID_JSON_BODY' || candidate.statusCode === 400;
}

function isFastifyRateLimitError(error: unknown): error is FastifyJsonError {
  if (!(error instanceof Error)) {
    return false;
  }

  const candidate = error as FastifyJsonError;
  return candidate.statusCode === 429 || candidate.code === 'FST_ERR_RATE_LIMIT';
}

export function mapToAppError(error: unknown): AppError {
  if (error instanceof AppError) {
    return error;
  }

  if (error instanceof ZodError) {
    return new ValidationError({
      detail: 'Request validation failed.',
      errors: error.issues,
      cause: error,
    });
  }

  if (isMalformedJsonError(error)) {
    return new ValidationError({
      detail: 'Malformed JSON body.',
      cause: error,
    });
  }

  if (isFastifyRateLimitError(error)) {
    return new RateLimitExceededError({
      detail: 'Too many requests, please try again later.',
      cause: error,
    });
  }

  if (error instanceof Error) {
    return new InternalServerError({
      detail: 'Unexpected server error.',
      cause: error,
      meta: {
        originalErrorName: error.name,
      },
    });
  }

  return new InternalServerError({
    detail: 'Unexpected server error.',
    meta: {
      unknownErrorType: typeof error,
    },
  });
}

File: flask-python-backend/src/errors/error_map.py
from werkzeug.exceptions import HTTPException

from .app_error import AppError
from .http_errors import InternalServerError, NotFoundError


def map_to_app_error(error: Exception) -> AppError:
    if isinstance(error, AppError):
        return error

    if isinstance(error, HTTPException):
        if error.code == 404:
            return NotFoundError(detail='Route not found.')

        status = error.code or 500
        title = error.name or 'HTTP Error'
        detail = error.description if isinstance(error.description, str) else 'Request failed.'
        return AppError(
            status=status,
            code=f'HTTP_{status}',
            title=title,
            detail=detail,
            is_operational=status < 500,
        )

    return InternalServerError()
